emcc

```diff
@@ -1340,14 +1344,42 @@ try:
 
   js_transform_tempfiles = [final]
 
-  if memory_init_file:
+  if memory_init_file or shared.Settings.MEM_INIT_METHOD == 2:
     memfile = target + '.mem'
     shared.try_delete(memfile)
     def repl(m):
       # handle chunking of the memory initializer
       s = m.groups(0)[0]
       if len(s) == 0 and not shared.Settings.EMTERPRETIFY: return m.group(0) # emterpreter must have a mem init file; otherwise, don't emit 0-size ones
-      open(memfile, 'wb').write(''.join(map(lambda x: chr(int(x or '0')), s.split(','))))
+      membytes = [int(x or '0') for x in s.split(',')]
+      if not shared.Settings.EMTERPRETIFY:
+        while membytes and membytes[-1] == 0:
+          membytes.pop()
+        if not membytes:
+          return '';
+      if not memory_init_file:
+        crcTable = []
+        for i in range(256):
+          crc = i
+          for bit in range(8):
+            crc = (crc >> 1) ^ ((crc & 1) * 0xedb88320)
+          crcTable.append(crc)
+        crc = 0xffffffff
+        s = list(membytes)
+        n = len(s)
+        crc = crcTable[(crc ^ n) & 0xff] ^ (crc >> 8)
+        crc = crcTable[(crc ^ (n >> 8)) & 0xff] ^ (crc >> 8)
+        for i in s:
+          crc = crcTable[(crc ^ i) & 0xff] ^ (crc >> 8)
+        for i in range(4):
+          s.append((crc >> (8 * i)) & 0xff)
+        s = ''.join(map(chr, s))
+        s = s.replace('\\', '\\\\').replace("'", "\\'")
+        s = s.replace('\n', '\\n').replace('\r', '\\r')
+        def escape(x): return '\\x{:02x}'.format(ord(x.group()))
+        s = re.sub('[\x80-\xff]', escape, s)
+        return "var memoryInitializer = '%s';" % s
```

Comment:
It took me a while to figure out that the call to `optimize_initializer` does that splitting, but only if we didn't use a memory initializer file or string. I could implement a similar splitting for the strings, in which case I'd probably turn the variable assignment into a function call, similar to the `alloc` call it's replacing. But perhaps we should only do so if it's really needed.

Investigating that need, do you have a one-off test in mind, or a part of the browser test suite? What browsers are officially supported? What kind of memory initializer size do you have in mind? Out of the blue, I'd try 32M and assume that if this works, any reasonable program will work.

Viewing the source could be hell anyway, as it currently stands, due to the non-printable control bytes included in the string literals. If that's a problem, we'll have to use way more hex escape sequences, and perhaps try to convince uglify to do something alike. I had the impression that small file sizes were more important here.